powerauth-restful-security-spring/src/main/java/io/getlime/security/powerauth/rest/api/spring/service/oidc/OAuth2Handler.java

```diff
@@ -0,0 +1,135 @@
+
+        final TokenRequest tokenRequest = TokenRequest.builder()
+                .clientId(oidcApplicationConfiguration.getClientId())
+                .code(request.getCode())
+                .clientSecret(oidcApplicationConfiguration.getClientSecret())
+                .clientAuthenticationMethod(oidcApplicationConfiguration.getClientAuthenticationMethod())
+                .tokenUrl(oidcApplicationConfiguration.getTokenUri())
+                .redirectUri(oidcApplicationConfiguration.getRedirectUri())
+                .build();
+
+        final TokenResponse tokenResponse = fetchToken(oidcApplicationConfiguration.getClientId(), tokenRequest);
+        final Jwt jwt = verifyAndDecode(tokenResponse, oidcApplicationConfiguration, request.getNonce());
+
+        final String subject = jwt.getSubject();
+        validateSubject(subject);
+        return subject;
+    }
+
+    private OidcApplicationConfiguration fetchOidcApplicationConfiguration(final OAuth2ActivationContext request) throws PowerAuthActivationException {
+        try {
+            return applicationConfigurationService.fetchOidcApplicationConfiguration(OidcConfigurationQuery.builder()
+                    .applicationKey(request.getApplicationKey())
+                    .providerId(request.getProviderId())
+                    .build());
+        } catch (PowerAuthApplicationConfigurationException e) {
+            throw new PowerAuthActivationException(e);
+        }
+    }
+
+    private TokenResponse fetchToken(final String clientId, final TokenRequest tokenRequest) throws PowerAuthActivationException {
+        logger.debug("Fetching token, clientId: {}", clientId);
+        try {
+            final TokenResponse response = tokenClient.fetchTokenResponse(tokenRequest);
+            logger.debug("Token fetched, verifying, clientId: {}", clientId);
+            return response;
+        } catch (RestClientException e) {
+            throw new PowerAuthActivationException("Unable to get token response", e);
+        }
+    }
+
+    private Jwt verifyAndDecode(final TokenResponse tokenResponse, final OidcApplicationConfiguration oidcApplicationConfiguration, final String nonce) throws PowerAuthActivationException {
+        final String issuerUrl = oidcApplicationConfiguration.getIssuerUri();
+
+        final NimbusJwtDecoder jwtDecoder = NimbusJwtDecoder.withIssuerLocation(issuerUrl)
+                .jwsAlgorithms(algorithms ->
+                        algorithms.addAll(List.of(
+                                SignatureAlgorithm.RS256,
+                                SignatureAlgorithm.RS384,
+                                SignatureAlgorithm.RS512,
+                                SignatureAlgorithm.ES256,
+                                SignatureAlgorithm.ES384,
+                                SignatureAlgorithm.ES512)))
+                .build();
+
+        final DelegatingOAuth2TokenValidator<Jwt> validator = new DelegatingOAuth2TokenValidator<>(
```

Comment:
Missing required validations:

- we do not decrypt encrypted token - it's OK for now
- The current time MUST be before the time represented by the `exp` Claim.

Missing recommended validation:
- we should store the nonce and check the uniqueness.  We can use validation of the `iat` Claim to limit the amount of time that nonces need to be stored to prevent attacks. 
- `alg` - should be RS256 or a specific algorithm.  

There are few other validations we may implement - `azp`, `acr`, `auth_time`, but it is not necessary as of now.

Reply:
`exp` is validated by default validator, see https://github.com/spring-projects/spring-security/blob/main/oauth2/oauth2-jose/src/main/java/org/springframework/security/oauth2/jwt/JwtValidators.java#L86

Reply:
Algorithms are listed here https://github.com/wultra/powerauth-restful-integration/pull/547/files#diff-aaa750e3df782f29e1aa7a7d51be596c1cc82caf61696ccd06001df58767f539R106
as 
```java
.jwsAlgorithms(algorithms ->
                        algorithms.addAll(List.of(
                                SignatureAlgorithm.RS256,
                                SignatureAlgorithm.RS384,
                                SignatureAlgorithm.RS512,
                                SignatureAlgorithm.ES256,
                                SignatureAlgorithm.ES384,
                                SignatureAlgorithm.ES512)))
```

Reply:
`nonce` is validated here https://github.com/wultra/powerauth-restful-integration/pull/547/files#diff-21cc1915f56f17444335ca51930a2421f05212af9e5d7531d522648a7102c678R57
But let's review it again when I attempt to switch to Spring `oauth-client`.

Reply:
> Algorithms are listed here https://github.com/wultra/powerauth-restful-integration/pull/547/files#diff-aaa750e3df782f29e1aa7a7d51be596c1cc82caf61696ccd06001df58767f539R106 as
> 
> ```java
> .jwsAlgorithms(algorithms ->
>                         algorithms.addAll(List.of(
>                                 SignatureAlgorithm.RS256,
>                                 SignatureAlgorithm.RS384,
>                                 SignatureAlgorithm.RS512,
>                                 SignatureAlgorithm.ES256,
>                                 SignatureAlgorithm.ES384,
>                                 SignatureAlgorithm.ES512)))
> ```

We are supporting implicitly set of algorithms. The correct implementation is:
Implicitly use RS256. If other alghoritm is explicitly agreed with the issuer support the agreed alghoritm.

Reply:
Yes, the Spring implementation is forcing a single algorithm, RS256 by default.